### ISSUE TITLE
Auto switch based on IP

### ISSUE THREAD

**Issue by fwlx** [FelisCatus/SwitchyOmega#197]
This is an old request which used to be raised on Switchy project in code.google.com
Resubmit again in case someone could add it in SwitchyOmega

This feature is indeed helpful when switch network between office/home
original request
https://code.google.com/p/switchy/issues/detail?id=41&colspec=ID%20Type%20Status%20Pri%20Mstone%20Summary%20Stars%20Reporter

**Comment by FelisCatus** [comment]
I think this is possible with `myIpAddress()` in PAC script. But I don't know how to get the local IP in Chrome Extensions. Is it acceptable if the proxy is switched correctly (using PAC), but the color of the icon is wrong (due to lack of local ip in Chrome Extension)?

Also, the `myIpAddress()` is not stable could give incorrect or meaningless results (like `127.0.0.1`). Which proxy should be used if `myIpAddress()` fails?

**Comment by in-out** [comment]
it works for me
var p1= "PROXY 127.0.0.1:8087;"
var p2 = "SOCKS5 127.0.0.1:1080; SOCKS 127.0.0.1:1080;";

function FindProxyForURL(url, host) {
    if (shExpMatch(myIpAddress(), "192.*")){
        return p1;
    }
    else{
        return p2;
    }
}

**Comment by FelisCatus** [comment]
@in-out Yes, it definitely works in PAC scripts. The main focus of this issue is to achieve the same thing in Chrome Extensions to keep the user interface consistent. The situation of `myIPAddress()` failing is an edge case in my option, but I just added that here for the record.

**Comment by grimblefritz** [comment]
I understand the limitation, that extensions cannot get the IP address of the local machine. However, the only real annoyance I have with SO (or any of the Chrome proxy tools) is that I work from home and the only time I need a proxy is when I'm connected to a VPN for work. If not for some of those VPNs also requiring authentication, I have a PAC file that works perfectly.

I propose this work around. As an auto-switch rule, allow me to enter an IP address that I deem to be "local". If SO can open a connection to that address, then I am using the local net and need no proxy; otherwise, if the connection fails, then I am using a VPN and need a proxy.

More generically, this could be a "if host (name or ip) exists, then ..." rule.

I recognize this could be slow; however, I for one would be willing to put up with it if it is not terribly slow.

**Comment by FelisCatus** [comment]
Connecting to an IP address is neither possible in Chrome Extensions or PAC scripts. It is possible to send UDP packets (and thus ping packets) in Package Apps but not Extensions. AFAIK, Chrome extensions is limited to sending HTTP requests, which could work only if you have a local HTTP server running on your own laptop.

So the situation would be: Proxy is switched correctly (using PAC), but the color of the icon is wrong (due to lack of knowledge of local IP in Chrome Extension), **unless you have a HTTP server listening on your own device and bound to a given address.** It is an improvement of detection, but more inconsistent in my opinion.

I'd rather say "connecting" a server won't help here. It is sad to say so, we have to bare with the wrong icon color unless we find another approach. To be fair, this works just like the @in-out 's solution: The actual behavior is correct, while you can't visually validate it through the user interface.

**Comment by pmaddaloni** [comment]
I'm sorry, but I'm a bit confused here; is this possible to do? I'd like use my proxy when I'm not on my home network, and ignore it when I am on my home network. I see that @in-out wrote a short script above, but where does that script go?

**Comment by FelisCatus** [comment]
It is technically possible to switch between proxy servers based on IP. However, extensions cannot know which server is actually being used and therefore cannot give you any feedback like changing the color of the icon.

The script by @in-out does work if put into a newly created PAC Profile and can cover your use case as intended. But the same restrictions applies and thus the extension icon might be misleading, which is the only downside.

**Comment by FelisCatus** [comment]
This is not really related to the issue, but I will post it here in case anyone need.

If you want use some proxy server only if it is online, you can use the following script for PAC Profile.

```js
function FindProxyForURL(url, host) {
    return "PROXY 111.11.111.111; DIRECT";
}
```

What it does is try connecting to `111.11.111.111` first. If the proxy server is not reachable (probably because you are not at the company), it falls back to direct connection. This should cover most of the cases in my opinion.

**Comment by FelisCatus** [comment]
The WebRTC leak exposes all local IP addresses. However, `myIpAddress` returns only one of them and the selection is not fully predictable. It may also return `127.0.0.1` instead of any useful information in some cases. The problem is still there: You can't make the PAC script and SwitchyOmega agree with each other. Therefore, I don't think we should go with the WebRTC solution.

P.S. No, WebRTC doesn't work in PAC scripts.

**Comment by FelisCatus** [comment]
Using the `myIpAddressEx` extension in PAC, we could probably create a more reliable type of condition.

**Description:** Matches when any of the local IP addresses is in range.
**Params:** IP range (maybe using CIDR notation)
**Implementation (PAC):** Use myIpAddressEx to get all IP addresses, then `isInNet` or `isInNetEx` to see any of them are in range or not. Return true if any of them is in range.
**Implementation (Chrome Extension):** Use the WebRTC leak to list all local IP addresses and compare them to ranges using the `ip-address` library. Return true if any of them is in range.

This is more likely to work because both sides are listing all IP addresses. There could still be differences between these two, but I don't think that matters much. How do you like these conditions? Any feedback?

**Comment by jan-glx** [comment]
I have difficulties to get `myIPAddressEx` to work with Google Chrome Version 65.0.3325.146 (Official Build) (64-bit). Does it work for you? How do you use it?

**Comment by niawag** [comment]
Hi, is this issue still a thing ? It's a really interesting feature for people who work from home/work.
